### PR TITLE
TC-IDM-10.1: add checks for no unexpected global/standard range attributes

### LINKED ISSUE #28287: The test plan has "SHALL NOT: bits that are not reflected in this YAML.  File a followup issue, please, add a comment in the YAML pointing to it.
Many .yaml test cases don't seem to account for the "SHALL NOT" include values when testing AttributeList, EventList, AcceptedCommandList, and GeneratedCommandList.

_Originally posted by @bzbarsky-apple in https://github.com/project-chip/connectedhomeip/pull/28023#discussion_r1274099041_

### PR DESCRIPTION
**Pull request by cecille**
As discussed on slack, we can do the last part of all the 1.1 checks globally and remove the manual piece from the 1.1 tests.

Test: changed the test to use the attribute list as the source of truth (this is reconciled with the returned attributes later in the test). Then added two fake values (1 global, 1 standard) into the attribute list at the point of encoding. Test fails as expected.

### CODE REVIEW

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (tcarmelveilleux) [inline src/python_testing/TC_DeviceBasicComposition.py:345]: Can we use a symbolic name for 0xFFFA to make it clear beyond the comment as well

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (tcarmelveilleux) [inline src/python_testing/TC_DeviceBasicComposition.py:351]:
```suggestion
                unexpected_globals = sorted(list(set(globals) - set(allowed_globals)))
```

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (tcarmelveilleux) [inline src/python_testing/TC_DeviceBasicComposition.py:362]:
```suggestion
                    # Skip clusters that are not part of the standard generated corpus (e.g. MS clusters)
                    continue
```

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (tcarmelveilleux) [inline src/python_testing/TC_DeviceBasicComposition.py:365]:
```suggestion
                unexpected_standard_attributes = sorted(list(set(standard_attributes) - set(allowed_standard_attributes)))
```

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (tcarmelveilleux) [inline src/python_testing/TC_DeviceBasicComposition.py:369]:
```suggestion
                                      problem=f"Unexpected standard attribute {unexpected} (0x{unexpected:04x}) in cluster {cluster_id} (0x{cluster_id:04x})", spec_location=f"Cluster {cluster_id} (0x{cluster_id:04x})")
```

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:345]: sure - will add.

Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py]: added

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (bzbarsky-apple) [inline src/python_testing/TC_DeviceBasicComposition.py:366]: This isn't right.  The top of the non-global standard attribute range is 0x4FFF, not 0xEFFF, so that test should be `a < 0x5000` or `a <= 0x4FFF` or something like that, perhaps with better naming.

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (bzbarsky-apple) [inline src/python_testing/TC_DeviceBasicComposition.py:369]: Does the test plan have checks for manufacturer-scoped invalid values?  i.e. MEIs where the prefix is not 0 and the suffix is larger than 0x4FFF (so not a non-global attribute) but less than 0xF000 (so not a global attribute)?

If not, it really should, and we should check for it here.

Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:369]: no, but I don't think that's specified in the spec either. table 70 says the valid range of suffixes for mei is 0x0000 to 0xfffe

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:366]: It ends up being the same thing because the spec will never define a standard value in the range above 0x4FFF, but I can make the check separate to make that obvious

Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py]: separated.

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (bzbarsky-apple) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: @cecille You want to look at table 71, which has different allowed suffix ranges based on type for data model types.

Comment (bzbarsky-apple) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: Also, table 71 disallows MC global attributes, as far as I can tell....

Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: This isn't covered in the current test plan. We can consider adding it in a follow up if you open an issue but I want to be really sure of the requirements before we add this to the cert test.

Comment (ReneJosefsen) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: The current test plan states:

```
The list SHALL NOT contain any additional values in the standard or scoped range: (0x0000_0000 - 0x0000_4FFF) and (0x0000_F000 - 0x0000_FFFE)

The list MAY contain values in the Manufacturer Extensible Identifier (MEI) range: (0xXXXX_0000 - 0xXXXX_4FFF), where XXXX is the allowed MEI range (0x0001 - 0xFFF1), these values SHALL be ignored.

The list SHALL NOT contain any values in the Test Vendor or invalid range: (0x0000_5000 - 0x0000_EFFF and 0x0000_FFFF), (0xXXXX_5000 - 0xXXXX_FFFF) and (0xFFF1_0000 - 0xFFFF_FFFF), where XXXX is the allowed MEI range (0x0001 - 0xFFF1)
```

So the test plan does have this ranges listed as not allowed?

Comment (ReneJosefsen) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: As far as I understand this test step checks this part of the statement:

**The list SHALL NOT contain any additional values in the standard or scoped range: (0x0000_0000 - 0x0000_4FFF) and (0x0000_F000 - 0x0000_FFFE)**

Comment (ReneJosefsen) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: I think there is a mistake in the allowed MEI range stated by the Test plan: (0x0001 - 0xFFF1), as far as I can see it should be 0x0001 - 0xFFF0.

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (ReneJosefsen) [inline src/python_testing/TC_DeviceBasicComposition.py:350]: I assume we could just use ATTRIBUTE_LIST_ID here?

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (tcarmelveilleux) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: Re: MEI list: FFF1 is used for SDK testing, and is present in all-clusters app at least. Breaking on FFF1 means we cannot add test items from SDK

Comment (ReneJosefsen) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: The Test Vendor MC is defined as 0xFFF1 - 0xFFF4, but my assumption is that these are not allowed on production devices? 🤔 

Manufacturer Code is 0x0001 - 0xFFF0, so either we allow 0x0001 - 0xFFF4 or 0x0001 - 0xFFF0?

Comment (bzbarsky-apple) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: > We can consider adding it in a follow up if you open an issue but I want to be really sure of the requirements before we add this to the cert test.

Looks like the test plan says this already.  I am fine with a separate PR to handle this part, @cecille, but I don't think it needs test plan changes.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: The current test plan doesn't disallow MEI in the gloabl range, though, right? Re: table 71 - are you saying global is disallowed in the MEI range because it's not listed as a permitted source? Is "MC" in that table manufacturer... something? cluster?

This is very confusing....I'm looking right at the spec and still not totally sure what's allowed and not. Or why...why restrict the MEI range?

Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: OK, let's address this in a follow up and get this PR in. I'm not technically supposed to be here, and I want to make sure we're all on the same page with MEI before it gets added. IMO the MEI range is less concerning than the standard range, so I'd like to get in what we can before SVE.

issue: https://github.com/project-chip/connectedhomeip/issues/28945

Comment (ReneJosefsen) [reply src/python_testing/TC_DeviceBasicComposition.py:370]: The way I interpret https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Data-Model-Namespace.adoc#ref_mei_data_model_types is that **Attribute ID (Global)** is only allowed for the Standard range, which is also why the TP only states 0xXXXX_0000 - 0xXXXX_4FFF as the allowed range for MEI range, I guess this is called MC in spec so we might want to align on that.

The latter sentence does includes (0xXXXX_5000 - 0xXXXX_FFFF) as a disallowed section.

Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:445]: ok ,what's the final call here, then? Add a check for no values in the MEI prefix range above the standard suffix range? and allow prefix values in the test range?

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py]: thanks, yeah. Good call - changed.

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (ReneJosefsen) [reply src/python_testing/TC_DeviceBasicComposition.py:444]: https://github.com/project-chip/connectedhomeip/pull/28926 does the checks for no attributes in (0x0000_5000 - 0x0000_EFFF and 0x0000_FFFF), (0xXXXX_5000 - 0xXXXX_FFFF) and (0xFFF1_0000 - 0xFFFF_FFFF), so once this merges we can move it into this test case?

Comment (bzbarsky-apple) [reply src/python_testing/TC_DeviceBasicComposition.py:444]: @ReneJosefsen OK to resolve this conversation then, since that test will handle the checks I asked for above?

Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:444]: I added it to this PR so I could get this convo resolved and get it in. PTAL

Comment (ReneJosefsen) [reply src/python_testing/TC_DeviceBasicComposition.py:444]: Sure, then we need to carve that part out of #28926 when we move it into this script.

**Review by ReneJosefsen — APPROVED** [review]

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (ReneJosefsen) [inline src/python_testing/TC_DeviceBasicComposition.py:460]: This is mostly what the script I have put up in TC_XX_1_1 is doing as well 😅 so there is getting quite some overlap now in regards to merging these two test script together once this merges.

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:444]: Ok, this is fine for now, resolving so we can merge

File: src/python_testing/TC_DeviceBasicComposition.py
Comment (cecille) [reply src/python_testing/TC_DeviceBasicComposition.py:460]: As per slack discussion we will dedupe these later.

**Review by tima-q — APPROVED** [review]

**Review by tcarmelveilleux — COMMENTED** [review]
LGTM

**Pull request merged**
